arangod/Aql/OptimizerRules.cpp

```diff
@@ -3903,13 +3903,261 @@ void arangodb::aql::scatterInClusterRule(Optimizer* opt, std::unique_ptr<Executi
+    case ExecutionNode::SHORTEST_PATH: {
+      auto shortestPathNode = ExecutionNode::castTo<ShortestPathNode const*>(node);
+      TRI_ASSERT(shortestPathNode->isDisjoint());
+      collection = shortestPathNode->collection();
+      inputVariable = shortestPathNode->startInVariable();
+      // TODO:
+      // If the traversal node uses a constant start vertex, then this will be
+      // nullptr, hence we'll have to stunt around this.
+      TRI_ASSERT(inputVariable);
+      alternativeVariable = inputVariable;
+      allowSpecifiedKeys = true;
+      createKeys = false;
+    } break;
+    default: {
+      TRI_ASSERT(false);
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                     "Cannot distribute " +
+                                         node->getTypeString() + ".");
+    } break;
+  }
+  TRI_ASSERT(collection != nullptr);
+  auto distNode =
+      plan.createNode<DistributeNode>(&plan, plan.nextId(), ScatterNode::ScatterType::SHARD,
+                                      collection, inputVariable, alternativeVariable,
+                                      createKeys, allowSpecifiedKeys);
+  TRI_ASSERT(distNode != nullptr);
+  return distNode;
+}
+
+// Create a new GatherNode for the DistributeNode passed in node, and
+// register it with the plan
+//
+// TODO: Really Scatter/Gather and Distribute/Gather should be created in pairs.
+auto arangodb::aql::createGatherNodeFor(ExecutionPlan& plan, DistributeNode* node)
+    -> GatherNode* {
+  auto const collection = node->collection();
+
+  auto const sortMode = GatherNode::evaluateSortMode(collection->numberOfShards());
+  auto const parallelism = GatherNode::Parallelism::Undefined;
+  return plan.createNode<GatherNode>(&plan, plan.nextId(), sortMode, parallelism);
+}
+
+//
+// for a node `at` of type
+//  - INSERT, REMOVE, UPDATE, REPLACE, UPSERT
+//  - TRAVERSAL, SHORTEST_PATH, K_SHORTEST_PATHS,
+// we transform
+//
+// parents[0] -> `node` -> deps[0]
+//
+// into
+//
+// parents[0] -> GATHER -> REMOTE -> `node` -> REMOTE -> DISTRIBUTE -> deps[0]
+//
+// We can only handle the above mentioned node types, because the setup of
+// distribute and gather requires knowledge from these nodes.
+//
+// Note that parents[0] might be `nullptr` if `node` is the root of the plan,
+// and we handle this case in here as well by resetting the root to the
+// inserted GATHER node.
+//
+auto arangodb::aql::insertDistributeGatherSnippet(ExecutionPlan& plan,
+                                                  ExecutionNode* at, SubqueryNode* snode)
+    -> DistributeNode* {
+  auto const parents = at->getParents();
+  auto const deps = at->getDependencies();
+
+  // This transforms `parents[0] -> node -> deps[0]` into `parents[0] -> deps[0]`
+  plan.unlinkNode(at, true);
+
+  // create, and register a distribute node
+  DistributeNode* distNode = createDistributeNodeFor(plan, at);
+  TRI_ASSERT(distNode != nullptr);
+  distNode->addDependency(deps[0]);
```

Comment:
```suggestion
TRI_ASSERT(deps.size() == 1);
  distNode->addDependency(deps[0]);
```

I think this code is only valid if you have EXACTLY one dependency.

I am rather certain that we right now only create those plans at this place, however this assertion helps in the future.